### PR TITLE
DAOS-13293 cart: add APIs to query protocol info from mercury

### PR DESCRIPTION
**Pull request by soumagne**
- Add crt_hg_get_protocol_info()/crt_hg_free_protocol_info()
- This will allow control plane to collect fabric interfaces and their supported providers.

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by kjacque — APPROVED** [review]
LGTM. Does this fetch both libfabric and ucx information? I guess it depends on the version of mercury installed?

**Comment by soumagne** [comment]
@kjacque yes it will work for both ucx and ofi. If you want everything you can pass NULL in the `info_string` but if you pass `ofi+tcp` for example it would fetch only for ofi with tcp, same for ucx etc. So you can use the `info_string` as a way of restricting which plugin/protocol you query. To give you more examples, if you pass `tcp` it would return the info for both ofi and ucx but only related to the tcp protocol. There is also a new utility with latest mercury called `hg_info` that you can use to experiment with and see what gets returned. You can try `hg_info tcp`, `hg_info ofi+tcp` or `hg_info` without any other arguments which is equivalent to passing NULL to the info_string.

File: src/cart/crt_hg.h
Comment (frostedcmos) [inline src/cart/crt_hg.h:101]: these are not exposed to the rest of the daos. we need to have external functions in src/include/cart/api.h, something like crt_get_protocol_info()  / crt_free_protocol_info()

**Review by frostedcmos — CHANGES_REQUESTED** [review]

File: src/include/cart/types.h
Comment (frostedcmos) [inline src/include/cart/types.h:27]: by our convention we typedef external struct types to 'struct_name'_t  so  crt_protocol_info_t

File: src/tests/ftest/cart/protocol_info.c
Comment (frostedcmos) [inline src/tests/ftest/cart/protocol_info.c:2]: copyright date update needed

File: src/tests/ftest/cart/protocol_info.c
Comment (frostedcmos) [inline src/tests/ftest/cart/protocol_info.c:6]: would be good to have a comment here describing the purpose or what this test does

File: src/tests/ftest/cart/protocol_info.c
Comment (frostedcmos) [inline src/tests/ftest/cart/protocol_info.c:22]: use D_ERROR() for these, so that all tests can end up writing to the same log files in CI

File: src/tests/ftest/cart/protocol_info.c
Comment (frostedcmos) [inline src/tests/ftest/cart/protocol_info.c:48]: either crt_init() or d_log_init() here (with a corresponding fini at the end) to get the logging capabilities. almost all of our tests use gurt level macros to print (D_ERROR/D_WARN/D_INFO/D_DEBUG). Makes it easier to find logs in the proper places in CI later.

We also have DBG_PRINT() macro as part of crt_utils which duplicates to cart logs and stderr/stdout, just need to call this function then to initialize it:

 void crtu_test_init(d_rank_t rank, int num_attach_retries, bool is_server, bool assert_on_error);

**Review by frostedcmos — CHANGES_REQUESTED** [review]
few comments inline.

Since you already adding this as ftest we also need to add python wrapper with yaml file.

This can be used as a basic sample to duplciate where you can take only client-side portion:
src/tests/ftest/cart/ctl/ctl_one_node.py
src/tests/ftest/cart/ctl/ctl_one_node.yaml

File: src/include/cart/api.h
Comment (ashleypittman) [inline src/include/cart/api.h:55]:
```suggestion
 * \param[in,out] protocol_info linked-list of protocol infos
```

File: src/include/cart/types.h
Comment (soumagne) [reply src/include/cart/types.h:27]: It's in the C style guide to not use typedefs actually:
> Do not use typedef for new declarations, as this hides the type of a field for very little benefit. Never typedef pointers, as the * makes C pointer declarations obvious. Hiding it inside a typedef just obfuscates the code.

Comment (soumagne) [reply src/include/cart/types.h:27]: @ashleypittman since you are the main maintainer of the style guide, can you please confirm? :)

Comment (ashleypittman) [reply src/include/cart/types.h:27]: I can confirm that a large part of our public API is in contradiction to our style guidelines.

Personally my preferred style is what you've done here as I find that typedefs only serve to obfuscate which is exactly what the coding style says and I didn't even write that part.

File: src/cart/crt_init.c
Comment (frostedcmos) [inline src/cart/crt_init.c:566]: _Static_assert is more appropriate here:)

Comment (soumagne) [reply src/cart/crt_init.c:566]: ah ok sure I can use that

**Review by frostedcmos — DISMISSED** [review]

**Comment by soumagne** [comment]
@kjacque once this is merged, have a look at the test in utest/utest_protocol.c for an example

File: utils/rpms/daos.spec
Comment (frostedcmos) [inline utils/rpms/daos.spec:370]: why not utest_*.c ? its fine as is, just curious:)

**Review by frostedcmos — DISMISSED** [review]

File: utils/rpms/daos.spec
Comment (soumagne) [reply utils/rpms/daos.spec:370]: no idea but I just thought I'd stick to what was previously done :) @brianjmurrell might know :)

File: src/tests/ftest/cart/utest/utest_protocol.c
Comment (jolivier23) [inline src/tests/ftest/cart/utest/utest_protocol.c:15]: You should add this test to utils/utest.yaml and ci/test_files_to_stash.txt

See https://github.com/daos-stack/daos/pull/12993

**Review by jolivier23 — CHANGES_REQUESTED** [review]

File: src/tests/ftest/cart/utest/utest_protocol.c
Comment (jolivier23) [reply src/tests/ftest/cart/utest/utest_protocol.c:15]: You may also want to run utils/run_utest.py --test_filter=protocol --memcheck to ensure it's valgrind clean before submitting though

Comment (soumagne) [reply src/tests/ftest/cart/utest/utest_protocol.c:15]: actually looks like valgrind is complaining about something, have to look at it...

Comment (soumagne) [reply src/tests/ftest/cart/utest/utest_protocol.c:15]: leak is fixed in https://github.com/mercury-hpc/mercury/pull/706 but will have to update mercury to pick it up, disabling part of that test in the meantime

Comment (ashleypittman) [reply src/tests/ftest/cart/utest/utest_protocol.c:15]: I'd expect we're updating anyway so perhaps this PR would be best waiting until that upgrade has landed, then we can land this one with all tests enabled.

Comment (soumagne) [reply src/tests/ftest/cart/utest/utest_protocol.c:15]: I'm not planning to push a new mercury update right at the moment so I would still prefer to get this merged so that at least @kjacque has a chance to experiment with the refactoring she wanted to do using those APIs. (And I would not expect her to need to pass the type of string that was causing a leak as she would just most certainly pass `NULL` as the input argument to get the list of all available transports)

File: src/tests/ftest/cart/utest/utest_protocol.c
Comment (frostedcmos) [inline src/tests/ftest/cart/utest/utest_protocol.c:66]: I am surprised our checkpatch didnt complain about // comments

**Review by frostedcmos — APPROVED** [review]

File: utils/rpms/daos.spec
Comment (brianjmurrell) [reply utils/rpms/daos.spec:370]: I have no objection to `utest_*.c`.  I was just being conservative an only removing what I knew should not be in there and not wanting to create some issue down the road with removing more than what should be there.

**Review by jgmoore-or — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

File: src/tests/ftest/cart/utest/utest_protocol.c
Comment (jolivier23) [reply src/tests/ftest/cart/utest/utest_protocol.c:66]: we use clang-format now.  It did complain about other stuff but I think it doesn't presently comment on code

Comment (ashleypittman) [reply src/tests/ftest/cart/utest/utest_protocol.c:66]: I was sure at one point we also used compiler option that threw errors on this as well, I'm surprised to see it was allowed.

**Pull request merged**
